File: Oryx.Cognite/src/Playground/Relationships.fs
// Copyright 2019 Cognite AS
// SPDX-License-Identifier: Apache-2.0

namespace Oryx.Cognite.Playground

open System.Net.Http

open Oryx
open Oryx.Cognite.Playground

open System.Collections.Generic
open CogniteSdk
open CogniteSdk.Types.Common

/// Various asset HTTP handlers.

[<RequireQualifiedAccess>]
module Relationships =
    [<Literal>]
    let Url = "/relationships"

    /// <summary>
    /// Retrieves list of relationships matching filter, and a cursor if given limit is exceeded
    /// </summary>
    /// <param name="query">The query with limit and cursor.</param>
    /// <returns>List of relationships.</returns>
    let list (query: RelationshipQuery) : HttpHandler<HttpResponseMessage, ItemsWithCursor<Relationship>, 'a> =
        withLogMessage "Relationships:list"
        >=> listPlayground query Url

    /// <summary>
    /// Create new relationships in the given project.
    /// </summary>
    /// <param name="assets">The relationships to create.</param>
    /// <returns>List of created relationships.</returns>
    let create (items: RelationshipCreate seq) : HttpHandler<HttpResponseMessage, ItemsWithoutCursor<Relationship>, 'a> =
        withLogMessage "Relationships:create"
        >=> createPlayground items Url

    /// <summary>
    /// Delete multiple relationships in the same project.
    /// </summary>
    /// <param name="relationships">The list of externalIds for relationships to delete.</param>
    /// <returns>Empty result.</returns>
    let delete (externalIds: string seq) : HttpHandler<HttpResponseMessage, EmptyResponse, 'a> =
        let relationships = externalIds |> Seq.map Identity.Create
        let items = ItemsWithoutCursor(Items=relationships)
        withLogMessage "Relationships:delete"
        >=> deletePlayground items Url

    /// <summary>
    /// Retrieves information about multiple relationships in the same project. A maximum of 1000 relationships IDs may be listed per
    /// request and all of them must be unique.
    /// </summary>
    /// <param name="ids">The ids of the relationships to get.</param>
    /// <returns>Relationships with given ids.</returns>
    let retrieve (ids: string seq) : HttpHandler<HttpResponseMessage, ItemsWithoutCursor<Relationship>, 'a> =
        let relationships = ids |> Seq.map Identity.Create
        withLogMessage "Relationships:retrieve"
        >=> retrievePlayground relationships Url

    /// <summary>
    /// Retrieves a list of relationships matching the given criteria. This operation does not support pagination.
    /// </summary>
    /// <param name="query">relationships search query.</param>
    /// <returns>List of relationships matching given criteria.</returns>
    let search (query: RestrictedGraphQuery) : HttpHandler<HttpResponseMessage, RestrictedGraphQueryResult, 'a> =
        withLogMessage "Relationships:search"
        >=> searchPlayground query Url

File: CogniteSdk/src/Resources/Beta/Relationships.cs
// Copyright 2020 Cognite AS
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

using CogniteSdk.Beta;
using static Oryx.Cognite.Beta.HandlerModule;
using HttpContext = Oryx.Context<System.Net.Http.HttpResponseMessage>;

namespace CogniteSdk.Resources.Beta
{
    // TODO: Move this out of namespace? CogniteSdk.Beta namespace is ignored when outside of namespace for some reason
    using CogniteSdk.Beta;

    /// <summary>
    /// For internal use. Contains all relationship methods.
    /// </summary>
    public class RelationshipResource : Resource
    {
        /// <summary>
        /// Will only be instantiated by the client.
        /// </summary>
        /// <param name="authHandler">The authentication handler.</param>
        /// <param name="ctx">Context to use for the request.</param>
        internal RelationshipResource(Func<CancellationToken, Task<string>> authHandler, HttpContext ctx) : base(authHandler, ctx)
        {
        }

        /// <summary>
        /// Retrieves list of relationships matching query.
        /// </summary>
        /// <param name="query">The query filter to use.</param>
        /// <param name="token">Optional cancellation token to use.</param>
        /// <returns>List of Relationships matching given filters and optional cursor</returns>
        public async Task<ItemsWithCursor<Relationship>> ListAsync(RelationshipQuery query, CancellationToken token = default)
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = Oryx.Cognite.Beta.Relationships.list<ItemsWithCursor<Relationship>>(query);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Create Relationships.
        /// </summary>
        /// <param name="relationships">Relationships to create.</param>
        /// <param name="token">Optional cancellation token.</param>
        /// <returns>Sequence of created Relationships.</returns>
        public async Task<IEnumerable<Relationship>> CreateAsync(IEnumerable<RelationshipCreate> relationships, CancellationToken token = default)
        {
            if (relationships is null)
            {
                throw new ArgumentNullException(nameof(relationships));
            }

            var req = Oryx.Cognite.Beta.Relationships.create<IEnumerable<Relationship>>(relationships);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Delete multiple relationships in the same project, along with all their descendants in the relationship hierarchy if
        /// recursive is true.
        /// </summary>
        /// <param name="externalIds">The externalIds of relationships to delete.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> DeleteAsync(IEnumerable<string> externalIds, CancellationToken token = default)
        {
            if (externalIds is null)
            {
                throw new ArgumentNullException(nameof(externalIds));
            }

            var req = Oryx.Cognite.Beta.Relationships.delete<EmptyResponse>(externalIds);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieves information about multiple relationships in the same project. A maximum of 1000 relationships IDs may be listed
        /// per request and all of them must be unique.
        /// </summary>
        /// <param name="ids">The list of relationships identities to retrieve.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<IEnumerable<Relationship>> RetrieveAsync(IEnumerable<string> ids, CancellationToken token = default)
        {
            if (ids is null)
            {
                throw new ArgumentNullException(nameof(ids));
            }

            var req = Oryx.Cognite.Beta.Relationships.retrieve<IEnumerable<Relationship>>(ids);
            return await RunAsync(req, token).ConfigureAwait(false);
        }
    }
}